### PR TITLE
CVSB-3915

### PR DESCRIPTION
**Pull request by vladddragomir**
more tests to be added! don't worry 😝

### CODE REVIEW

**Comment by vladddragomir** [comment]
unit tests added!

File: src/pages/testing/test-creation/complete-test/complete-test.ts [hunk 1]
```diff
@@ -411,8 +424,35 @@ export class CompleteTestPage implements OnInit {
     this.cdRef.detectChanges();
     if (this.testTypeService.isTirTestType(this.vehicleTest.testTypeId)) {
       this.vehicleTest.certificateNumber = value.length > 5 ? value.substring(0, 5) : value;
+    } else if (this.testTypeService.isSpecialistTestTypesExceptForCoifAndVoluntaryIvaTestAndRetest(this.vehicleTest.testTypeId) ||
+      this.testTypeService.isSpecialistPartOfCoifTestTypes(this.vehicleTest.testTypeId) ||
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/complete-test/complete-test.ts]: This "A OR B OR C" logic is reused in several places. Could be consolidated to a single function.

File: src/pages/testing/test-creation/test-review/test-review.ts [hunk 2]
```diff
@@ -168,6 +185,8 @@ export class TestReviewPage implements OnInit {
         (!changedTestType.certificateNumber || (changedTestType.certificateNumber && changedTestType.certificateNumber.length < 5))) {
         this.popToTestCreatePage();
       }
+    } else if (!this.isSpecialistTestTypeCompleted(changedTestType, initialTestType)) {
```
Comment (KHodderDVSA) [inline src/pages/testing/test-creation/test-review/test-review.ts]: This whole function is a nightmare, that seems to have only one success output. Surely it can be simplified?

File: src/models/defects/defect-details.model.ts [hunk 3]
```diff
@@ -44,4 +44,10 @@ export interface DefectLocationModel {
   axleNumber?: number | null;
 }
 
-
+export interface SpecialistCustomDefectModel {
+  referenceNumber: string;
+  defectName: string;
+  defectNotes: string;
+  /* -------- ONLY FOR FE -------- */
+  hasAllMandatoryFields: boolean;
```
Comment (agoncearenco) [inline src/models/defects/defect-details.model.ts]: Wouldn't it be easier to just have a function that checks this? You won't have to worry of updating this flag every time you change a field value

File: src/models/tests/test-type.model.ts [hunk 4]
```diff
@@ -1,10 +1,11 @@
-import { DefectDetailsModel } from "../defects/defect-details.model";
+import { DefectDetailsModel, SpecialistCustomDefectModel } from "../defects/defect-details.model";
 
 export interface TestTypeModel {
   testTypeName: string;
   name: string;
   testTypeId: string;
   certificateNumber: string | null;
+  secondaryCertificateNumber: string | null;
```
Comment (agoncearenco) [inline src/models/tests/test-type.model.ts]: The secondary certificate number is not displayed in the app?

File: src/models/defects/defect-details.model.ts (hunk 3 above)
Comment (vladddragomir) [reply src/models/defects/defect-details.model.ts]: it's easier this way to be honest...i just update the value of this property every time i leave defect-details-specialist-testing.ts and whenever i press review on test overview page i am checking that every hasAllMandatoryFields property on customDefects is true, if not, trigger the validations.

File: src/pages/testing/test-creation/complete-test/complete-test.ts (hunk 1 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.ts]: i left it like this because it's more readable by others because you see right on the spot which test types are being checked for....my opinion

File: src/pages/testing/test-creation/test-review/test-review.ts (hunk 2 above)
Comment (vladddragomir) [reply src/pages/testing/test-creation/test-review/test-review.ts]: same as above....i just think that it's more readable this way, because you see exactly which conditions you should meet for every test type category. i mean it can be simplified but i don't know if this makes our lives easier

File: src/models/tests/test-type.model.ts (hunk 4 above)
Comment (vladddragomir) [reply src/models/tests/test-type.model.ts]: it will be on the next ticket you will review (10219) 😋

Comment (agoncearenco) [reply src/models/tests/test-type.model.ts]: Ah ok!

File: src/models/defects/defect-details.model.ts (hunk 3 above)
Comment (agoncearenco) [reply src/models/defects/defect-details.model.ts]: Ok

File: src/assets/app-data/test-types-data/notifiable-alteration-test-types.data.ts [hunk 5]
```diff
@@ -2,4 +2,8 @@ export class NotifiableAlterationTestTypesData {
   public static get NotifiableAlterationTestTypesDataIds(): string[] {
     return ['47', '48'];
   }
+
+  public static get PsvNotifiableAlterationTestTypeDataIds(): string[] {
```
Comment (agoncearenco) [inline src/assets/app-data/test-types-data/notifiable-alteration-test-types.data.ts]: Why did you need to split the PSV notifiable alteration test? Is it a specialist test? Doesn't it have to be also in the `NotifiableAlterationTestTypesDataIds`?

File: src/assets/app-data/test-types-data/specialist-test-types.data.ts [hunk 6]
```diff
@@ -2,4 +2,16 @@ export class SpecialistTestTypesData {
   public static get SpecialistTestTypesIds(): string[] {
```
Comment (agoncearenco) [inline src/assets/app-data/test-types-data/specialist-test-types.data.ts]: to not have to keep this list updated, you might want to return `SpecialistIvaTestAndRetestTestTypeIds` + `SpecialistTestTypesExceptForCoifAndVoluntaryIvaTestAndRetestIds` + `SpecialistPartOfCoifTestTypesIds` instead of a completely new list

Comment (vladddragomir) [reply src/assets/app-data/test-types-data/specialist-test-types.data.ts]: combining all these won't get the complete list of specialist test types

File: src/assets/app-data/test-types-data/notifiable-alteration-test-types.data.ts (hunk 5 above)
Comment (vladddragomir) [reply src/assets/app-data/test-types-data/notifiable-alteration-test-types.data.ts]: there is one ac on this ticket that adds an optional field only for this notifiable alteration test type which is for PSVs (the others are for hgvt). i didn't add it in that list because there is no bug against it and nobody said anything, i mean if it were to be an issue, someone would have raised it.

File: src/pages/testing/test-creation/complete-test/complete-test.html [hunk 7]
```diff
@@ -172,6 +173,28 @@ <h4 class="parent-info-color truncate-text" *ngIf="defect.deficiencyCategory !=
...
     </ion-item-group>
 
+    <ion-item-group *ngIf="testTypeDetails.hasSpecialistDefects">
+      <ion-item-divider text-uppercase>
+        <span>DEFECTS</span>
+      </ion-item-divider>
+      <ion-item-sliding #item *ngFor="let defect of vehicleTest.customDefects; let index = index">
+        <button ion-item class="defect-details custom-defects-divider"
+                [ngClass]="{'error-incomplete-item': errorIncomplete && !defect.hasAllMandatoryFields}"
+                (click)="toSpecialistDefectDetailsPage(true, index, defect)">
```
Comment (agoncearenco) [inline src/pages/testing/test-creation/complete-test/complete-test.html]: Why do you need both index in the `customDefects` array and the defect itself? Isn't this info redundant?

Comment (vladddragomir) [reply src/pages/testing/test-creation/complete-test/complete-test.html]: i have to be able to remove the custom defect from inside the custom-defect-details page. the index is the only unique property through which i can identify the specific defect
